File: shared/python/src/voiceai_contracts/session.py
"""Session — per-call working memory.

Mirrors shared/contracts/session.schema.json.
Stored in Redis as JSON under key `session:{call_uuid}` with TTL 2h.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


Role = Literal["user", "assistant"]


class Intent(str, Enum):
    VEHICLE_SEARCH = "vehicle_search"
    PRICING_QUERY = "pricing_query"
    AVAILABILITY_CHECK = "availability_check"
    FINANCING_INQUIRY = "financing_inquiry"
    DEALERSHIP_INFO = "dealership_info"
    APPOINTMENT_REQUEST = "appointment_request"
    HUMAN_TRANSFER = "human_transfer"


class SessionEntities(BaseModel):
    """Merged entities across conversation turns. Enables multi-turn refinement."""

    model_config = ConfigDict(extra="forbid")

    make: str | None = None
    model: str | None = None
    year_min: int | None = Field(default=None, ge=1900, le=2100)
    year_max: int | None = Field(default=None, ge=1900, le=2100)
    trim: str | None = None
    color_ext: str | None = None
    color_int: str | None = None
    body_style: str | None = None
    transmission: str | None = None
    fuel_type: str | None = None
    price_min: float | None = Field(default=None, ge=0)
    price_max: float | None = Field(default=None, ge=0)
    mileage_max: int | None = Field(default=None, ge=0)
    features: list[str] = Field(default_factory=list)
    condition: Literal["new", "used", "cpo"] | None = None


class ConversationTurn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: Role
    text: str
    ts: datetime
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class Session(BaseModel):
    """Per-call working memory stored in Redis."""

    model_config = ConfigDict(extra="forbid")

    call_uuid: str = Field(min_length=1)
    vonage_conversation_uuid: str | None = None
    caller_phone_hash: str = Field(min_length=64, max_length=64)
    caller_phone_last4: str | None = Field(default=None, pattern=r"^[0-9]{4}$")
    started_at: datetime
    ended_at: datetime | None = None
    entities: SessionEntities = Field(default_factory=SessionEntities)
    conversation_history: list[ConversationTurn] = Field(default_factory=list)
    tts_playing: bool = False
    intent: Intent | None = None
    transfer_attempts: int = Field(default=0, ge=0)

File: shared/typescript/src/session.ts
/**
 * Session — per-call working memory.
 * Mirrors ../contracts/session.schema.json.
 * Stored in Redis as JSON under key `session:{call_uuid}` with TTL 2h.
 */

import { z } from "zod";

// ─── Intent enum ───
export const IntentSchema = z.enum([
  "vehicle_search",
  "pricing_query",
  "availability_check",
  "financing_inquiry",
  "dealership_info",
  "appointment_request",
  "human_transfer",
]);
export type Intent = z.infer<typeof IntentSchema>;

// ─── Role ───
export const RoleSchema = z.enum(["user", "assistant"]);
export type Role = z.infer<typeof RoleSchema>;

// ─── ConditionFilter — distinct from Vehicle.Condition to allow `null` in filters ───
export const ConditionFilterSchema = z.enum(["new", "used", "cpo"]);
export type ConditionFilter = z.infer<typeof ConditionFilterSchema>;

// ─── SessionEntities ───
export const SessionEntitiesSchema = z
  .object({
    make: z.string().nullable().optional(),
    model: z.string().nullable().optional(),
    year_min: z.number().int().min(1900).max(2100).nullable().optional(),
    year_max: z.number().int().min(1900).max(2100).nullable().optional(),
    trim: z.string().nullable().optional(),
    color_ext: z.string().nullable().optional(),
    color_int: z.string().nullable().optional(),
    body_style: z.string().nullable().optional(),
    transmission: z.string().nullable().optional(),
    fuel_type: z.string().nullable().optional(),
    price_min: z.number().min(0).nullable().optional(),
    price_max: z.number().min(0).nullable().optional(),
    mileage_max: z.number().int().min(0).nullable().optional(),
    features: z.array(z.string()).default([]),
    condition: ConditionFilterSchema.nullable().optional(),
  })
  .strict();
export type SessionEntities = z.infer<typeof SessionEntitiesSchema>;

// ─── ConversationTurn ───
export const ConversationTurnSchema = z
  .object({
    role: RoleSchema,
    text: z.string(),
    ts: z.string().datetime({ offset: true }),
    confidence: z.number().min(0).max(1).nullable().optional(),
  })
  .strict();
export type ConversationTurn = z.infer<typeof ConversationTurnSchema>;

// ─── Session ───
export const SessionSchema = z
  .object({
    call_uuid: z.string().min(1),
    vonage_conversation_uuid: z.string().optional(),
    caller_phone_hash: z.string().length(64),
    caller_phone_last4: z
      .string()
      .regex(/^[0-9]{4}$/)
      .optional(),
    started_at: z.string().datetime({ offset: true }),
    ended_at: z.string().datetime({ offset: true }).nullable().optional(),
    entities: SessionEntitiesSchema,
    conversation_history: z.array(ConversationTurnSchema),
    tts_playing: z.boolean().default(false),
    intent: IntentSchema.nullable().optional(),
    transfer_attempts: z.number().int().min(0).default(0),
  })
  .strict();
export type Session = z.infer<typeof SessionSchema>;

/**
 * Helper: construct a fresh empty session for a new call.
 */
export function newSession(args: {
  call_uuid: string;
  caller_phone_hash: string;
  caller_phone_last4?: string;
  vonage_conversation_uuid?: string;
}): Session {
  return {
    call_uuid: args.call_uuid,
    vonage_conversation_uuid: args.vonage_conversation_uuid,
    caller_phone_hash: args.caller_phone_hash,
    caller_phone_last4: args.caller_phone_last4,
    started_at: new Date().toISOString(),
    ended_at: null,
    entities: { features: [] },
    conversation_history: [],
    tts_playing: false,
    intent: null,
    transfer_attempts: 0,
  };
}
